### PR TITLE
[LLVM][IR] Add constant range support for floating-point types

### PR DESCRIPTION
**Pull request by dtcxzyw**
Related issues: https://github.com/llvm/llvm-project/issues/68301 https://github.com/llvm/llvm-project/issues/70985 https://github.com/llvm/llvm-project/issues/82381

### CODE REVIEW

**Comment by dtcxzyw** [comment]
@tschuett This patch is not ready for review.

**Comment by tschuett** [comment]
https://github.com/llvm/llvm-project/blob/edcf65d40c4160db0a888e801560a671b10179d2/llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp#L6760

File: llvm/include/llvm/IR/ConstantFPRange.h [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+class raw_ostream;
+struct KnownFPClass;
+
+/// This class represents a range of floating-point values.
+class [[nodiscard]] ConstantFPRange {
+  APFloat Lower, Upper;
+  bool MaybeQNaN : 1;
+  bool MaybeSNaN : 1;
+  bool SignBitMaybeZero : 1;
+  bool SignBitMaybeOne : 1;
```
Comment (arsenm) [inline llvm/include/llvm/IR/ConstantFPRange.h:35]: Don't understand why these would be "MaybeZero" or "MaybeOne". Must be? Maybe would need to default to 1

File: llvm/include/llvm/IR/ConstantFPRange.h [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+public:
+  /// Initialize a full or empty set for the specified semantics.
+  explicit ConstantFPRange(const fltSemantics &FloatSema, bool IsFullSet);
+
+  /// Initialize a range to hold the single specified value.
+  ConstantFPRange(const APFloat &Value);
+
+  /// Initialize a range of values explicitly.
+  ConstantFPRange(APFloat Lower, APFloat Upper, bool MaybeQNaN, bool MaybeSNaN,
+                  bool SignBitMaybeZero, bool SignBitMaybeOne);
```
Comment (arsenm) [inline llvm/include/llvm/IR/ConstantFPRange.h:56]: Default values for the nan/sign params

File: llvm/include/llvm/IR/ConstantFPRange.h [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  }
+
+public:
+  /// Initialize a full or empty set for the specified semantics.
+  explicit ConstantFPRange(const fltSemantics &FloatSema, bool IsFullSet);
+
+  /// Initialize a range to hold the single specified value.
+  ConstantFPRange(const APFloat &Value);
+
+  /// Initialize a range of values explicitly.
```
Comment (jcranmer-intel) [inline llvm/include/llvm/IR/ConstantFPRange.h:54]: Might be worth having different methods for constructing an `(x, y)` range and a `[x, y]` range. Or maybe just a helper for `(-inf, inf)` (aka, all finite values); that seems to be the only common open range.

**Review by jcranmer-intel — COMMENTED** [review]
Not entirely clear to me from the UI whether or not the range is half-open or closed, which tends to matter more for floating-point values than it does for integers.

If it's half-open, then how is one supposed to represent `[inf, inf]`?

**Comment by dtcxzyw** [comment]
> Not entirely clear to me from the UI whether or not the range is half-open or closed, which tends to matter more for floating-point values than it does for integers.
> 
> If it's half-open, then how is one supposed to represent `[inf, inf]`?

Currently we use closed intervals and use `[inf,-inf]` as the canonical form of an empty set.
BTW, do we need wrapped sets to represent something like `{X: X < -10 || X > 10}`? If not, there is no way to implement `inverse`.

File: llvm/lib/IR/ConstantFPRange.cpp [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+// Represent a range of possible values that may occur when the program is run
+// for a floating-point value. This keeps track of a lower and upper bound for
+// the constant.
+//
+// Range = [Lower, Upper] U (MayBeQNaN ? QNaN : {}) U (MayBeSNaN ? SNaN : {})
+// Specifically, [inf, -inf] represents an empty set.
+// Note: -0 is considered to be less than 0. That is, range [0, 0] doesn't
+// contain -0.
+//
+//===----------------------------------------------------------------------===//
```
Comment (arsenm) [inline llvm/lib/IR/ConstantFPRange.cpp:18]: Comment belongs in the header, not the implementation

**Review by arsenm — COMMENTED** [review]
lgtm

**Comment by dtcxzyw** [comment]
> BTW, do we need wrapped sets to represent something like {X: X < -10 || X > 10}? If not, there is no way to implement inverse.

Any thoughts? IMO there will be a significant maintenance burden for us if we plan to support wrapped sets.

**Comment by arsenm** [comment]
> > BTW, do we need wrapped sets to represent something like {X: X < -10 || X > 10}? If not, there is no way to implement inverse.
> 
> Any thoughts? IMO there will be a significant maintenance burden for us if we plan to support wrapped sets.

I think it's potentially useful. Some instructions only operate in a range of values. Maybe doesn't need to be in the initial support

**Comment by nikic** [comment]
> > BTW, do we need wrapped sets to represent something like {X: X < -10 || X > 10}? If not, there is no way to implement inverse.
> 
> Any thoughts? IMO there will be a significant maintenance burden for us if we plan to support wrapped sets.

For integers, wrapped ranges are quite important because a non-wrapping signed range is often a wrapping unsigned range and vice versa. For FP, this problem does not exist as the ranges are always signed. I think it's better to leave out wrapped sets, at least for now, as they do make reasoning about operations a good bit harder.

File: llvm/include/llvm/IR/ConstantFPRange.h (hunk 3 above)
Comment (dtcxzyw) [reply llvm/include/llvm/IR/ConstantFPRange.h:62]: Done.

**Review by arsenm — APPROVED** [review]

File: llvm/unittests/IR/ConstantFPRangeTest.cpp
Comment (nikic) [inline llvm/unittests/IR/ConstantFPRangeTest.cpp:1]: We should have exhaustive testing infrastructure. On half, or one of the 8-bit types, if any of them is sufficiently IEEE compliant...

File: llvm/include/llvm/IR/ConstantFPRange.h [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+//===----------------------------------------------------------------------===//
+//
+// Represent a range of possible values that may occur when the program is run
+// for a floating-point value. This keeps track of a lower and upper bound for
+// the constant.
+//
+// Range = [Lower, Upper] U (MayBeQNaN ? QNaN : {}) U (MayBeSNaN ? SNaN : {})
+// Specifically, [inf, -inf] represents an empty set.
+// Note: -0 is considered to be less than 0. That is, range [0, 0] doesn't
+// contain -0.
```
Comment (nikic) [inline llvm/include/llvm/IR/ConstantFPRange.h:16]: This comment should explicitly say that the bounds are inclusive, and whether wrapping ranges are supported.

File: llvm/unittests/IR/ConstantFPRangeTest.cpp
Comment (arsenm) [reply llvm/unittests/IR/ConstantFPRangeTest.cpp:1]: I think Float8E4M3 covers everything

File: llvm/include/llvm/IR/ConstantFPRange.h (hunk 5 above)
Comment (dtcxzyw) [reply llvm/include/llvm/IR/ConstantFPRange.h]: Done.

File: llvm/unittests/IR/ConstantFPRangeTest.cpp
Comment (dtcxzyw) [reply llvm/unittests/IR/ConstantFPRangeTest.cpp:1]: `isSupportedSemantics` has been removed. I will add fcmp support with exhaustive testing infrastructure in follow-up patches.

File: llvm/lib/IR/ConstantFPRange.cpp [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+  return LHS.compare(RHS);
+}
+
+ConstantFPRange::ConstantFPRange(APFloat LowerVal, APFloat UpperVal,
+                                 bool MayBeQNaN, bool MayBeSNaN)
+    : Lower(std::move(LowerVal)), Upper(std::move(UpperVal)) {
+  // Canonicalize empty set into [Inf, -Inf].
+  if (strictCompare(Lower, Upper) == APFloat::cmpGreaterThan &&
+      !(Lower.isInfinity() && Upper.isInfinity()))
+    makeEmpty();
```
Comment (nikic) [inline llvm/lib/IR/ConstantFPRange.cpp:78]: Can we make this an assertion instead? Passing upper < lower seems more likely to be a bug than an attempt to create a non-canonical empty range.

File: llvm/unittests/IR/ConstantFPRangeTest.cpp
Comment (nikic) [reply llvm/unittests/IR/ConstantFPRangeTest.cpp:1]: We should also have exhaustive tests for classify at least (the correctness of the implementation is not obvious to me) and possibly also intersect/union.

File: llvm/unittests/IR/ConstantFPRangeTest.cpp [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+  EXPECT_FALSE(Full.isSingleElement());
+  EXPECT_FALSE(Empty.isSingleElement());
+  EXPECT_TRUE(One.isSingleElement());
+  EXPECT_FALSE(Some.isSingleElement());
+  EXPECT_FALSE(Zero.isSingleElement());
+}
+
+TEST_F(ConstantFPRangeTest, Enumerate) {
+  constexpr unsigned NNaNValues = (1 << 8) - 2 * ((1 << 3) - 1);
+  constexpr unsigned Expected = 4 * ((NNaNValues + 1) * NNaNValues / 2 + 1);
```
Comment (dtcxzyw) [inline llvm/unittests/IR/ConstantFPRangeTest.cpp:240]: We have 117616 ranges here. It is impossible to exhaustively test binary operations.

File: llvm/unittests/IR/ConstantFPRangeTest.cpp [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+
+  EXPECT_EQ(ToString(Full), "full-set");
+  EXPECT_EQ(ToString(Empty), "empty-set");
+  EXPECT_EQ(ToString(NaN), "NaN");
+  EXPECT_EQ(ToString(SNaN), "SNaN");
+  EXPECT_EQ(ToString(QNaN), "QNaN");
+  EXPECT_EQ(ToString(One), "[1, 1]");
+  EXPECT_EQ(ToString(Some.unionWith(SNaN)), "[-3, 3] with SNaN");
+}
+
```
Comment (arsenm) [inline llvm/unittests/IR/ConstantFPRangeTest.cpp:369]: Add a death test for the malformed ranges (e.g. lower bound greater than upper bound)?

Comment (dtcxzyw) [reply llvm/unittests/IR/ConstantFPRangeTest.cpp:369]: Done.

File: llvm/unittests/IR/ConstantFPRangeTest.cpp [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+    const APFloat &Upper = CR.getUpper();
+    if (!(Lower.isPosInfinity() && Upper.isNegInfinity())) {
+      while (true) {
+        Mask |= Lower.classify();
+        HasPos |= !Lower.isNegative();
+        HasNeg |= Lower.isNegative();
+        if (Lower.bitwiseIsEqual(Upper))
+          break;
+        strictNext(Lower);
+      }
```
Comment (nikic) [inline llvm/unittests/IR/ConstantFPRangeTest.cpp:341]: Add a helper to enumerate APFloats in ConstantFPRange?

File: llvm/unittests/IR/ConstantFPRangeTest.cpp (hunk 7 above)
Comment (nikic) [reply llvm/unittests/IR/ConstantFPRangeTest.cpp:240]: Hm yeah, I didn't think the numbers here through. Probably the best we can do instead is to have a list of interesting values (-inf, -largest_normal, -smallest_normal, -largest_subnormal, -smallest_subnormal, -0, +0, ...) plus a few normal numbers to use as boundary values. Not as good as full exhaustive coverage, but at least it makes sure that all the special cases are treated correctly.

**Review by arsenm — APPROVED** [review]

File: llvm/unittests/IR/ConstantFPRangeTest.cpp (hunk 7 above)
Comment (arsenm) [reply llvm/unittests/IR/ConstantFPRangeTest.cpp:240]: Should test with a bit in every mantissa position

File: llvm/unittests/IR/ConstantFPRangeTest.cpp (hunk 9 above)
Comment (dtcxzyw) [reply llvm/unittests/IR/ConstantFPRangeTest.cpp]: Done.

File: llvm/unittests/IR/ConstantFPRangeTest.cpp (hunk 7 above)
Comment (dtcxzyw) [reply llvm/unittests/IR/ConstantFPRangeTest.cpp:281]: Reduced to 2664 ranges.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by dtcxzyw** [comment]
@arsenm Is it ok to land this patch?

**Review by arsenm — APPROVED** [review]

**Comment by dtcxzyw** [comment]
> error: ignoring return value of function declared with 'nodiscard' attribute [-Werror,-Wunused-result]

**Comment by dtcxzyw** [comment]
> > error: ignoring return value of function declared with 'nodiscard' attribute [-Werror,-Wunused-result]

Fixed by https://github.com/llvm/llvm-project/pull/109921.

File: llvm/lib/IR/ConstantFPRange.cpp [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/IR/ConstantFPRange.h"
+#include "llvm/ADT/APFloat.h"
+#include "llvm/Analysis/ValueTracking.h"
```
Comment (chapuni) [inline llvm/lib/IR/ConstantFPRange.cpp:11]: `LLVMCore` should not depend on `llvm/Analysis`.

**Review by chapuni — COMMENTED** [review]
This has introduced a layering violation.

Pruning the problematic method may be a quick fix. Or please revert atm.

File: llvm/lib/IR/ConstantFPRange.cpp (hunk 10 above)
Comment (dtcxzyw) [reply llvm/lib/IR/ConstantFPRange.cpp:11]: Will post a fix later.

File: llvm/include/llvm/IR/ConstantFPRange.h [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+  ///
+  /// Example: Pred = ole and Other = float [2, 5] returns [-inf, 2]
+  static ConstantFPRange makeSatisfyingFCmpRegion(FCmpInst::Predicate Pred,
+                                                  const ConstantFPRange &Other);
+
+  /// Produce the exact range such that all values in the returned range satisfy
+  /// the given predicate with any value contained within Other. Formally, this
+  /// returns the exact answer when the superset of 'union over all y in Other
+  /// is exactly same as the subset of intersection over all y in Other.
+  /// { x : fcmp op x y is true}'.
```
Comment (jayfoad) [inline llvm/include/llvm/IR/ConstantFPRange.h:123]: The "Formally" part doesn't seem to make sense - did something go wrong with the single-quoted part? Isn't it just: `Formally, this returns { x : fcmp op x Other is true}`?

File: llvm/include/llvm/IR/ConstantFPRange.h [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+  /// Return false if the sign bit of all values in this range is 0.
+  /// Otherwise, return std::nullopt.
+  std::optional<bool> getSignBit() const;
+
+  /// Return true if this range is equal to another range.
+  bool operator==(const ConstantFPRange &CR) const;
+  /// Return true if this range is not equal to another range.
+  bool operator!=(const ConstantFPRange &CR) const { return !operator==(CR); }
+
+  /// Return the FPClassTest which will return true for the value.
```
Comment (jayfoad) [inline llvm/include/llvm/IR/ConstantFPRange.h:175]: This is a bit vague. Maybe "Return the smallest FPClassTest which includes all values in the range"?

File: llvm/include/llvm/IR/ConstantFPRange.h [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+  void print(raw_ostream &OS) const;
+
+  /// Allow printing from a debugger easily.
+  void dump() const;
+
+  /// Return the range that results from the intersection of this range with
+  /// another range.
+  ConstantFPRange intersectWith(const ConstantFPRange &CR) const;
+
+  /// Return the range that results from the union of this range
```
Comment (jayfoad) [inline llvm/include/llvm/IR/ConstantFPRange.h:191]: Maybe "Return the _smallest_ range ...", just to be clear?

File: llvm/include/llvm/IR/ConstantFPRange.h [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+  /// Return true if this set contains no members.
+  bool isEmptySet() const;
+
+  /// Return true if the specified value is in the set.
+  bool contains(const APFloat &Val) const;
+
+  /// Return true if the other range is a subset of this one.
+  bool contains(const ConstantFPRange &CR) const;
+
+  /// If this set contains a single element, return it, otherwise return null.
```
Comment (jayfoad) [inline llvm/include/llvm/IR/ConstantFPRange.h:159]: Maybe "a single non-NaN element" would be clearer, here and in `isSingleElement` below?

File: llvm/lib/IR/ConstantFPRange.cpp [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+}
+
+ConstantFPRange::ConstantFPRange(APFloat LowerVal, APFloat UpperVal,
+                                 bool MayBeQNaN, bool MayBeSNaN)
+    : Lower(std::move(LowerVal)), Upper(std::move(UpperVal)) {
+  assert(&Lower.getSemantics() == &Upper.getSemantics() &&
+         "Should only use the same semantics");
+  assert(!isNonCanonicalEmptySet(Lower, Upper) && "Non-canonical form");
+  this->MayBeQNaN = MayBeQNaN;
+  this->MayBeSNaN = MayBeSNaN;
```
Comment (jayfoad) [inline llvm/lib/IR/ConstantFPRange.cpp:91]: Can you not initialize these in the member initializer list?

File: llvm/lib/IR/ConstantFPRange.cpp [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+    Mask |= fcSNan;
+  if (MayBeQNaN)
+    Mask |= fcQNan;
+  if (!isNaNOnly()) {
+    FPClassTest LowerMask = Lower.classify();
+    FPClassTest UpperMask = Upper.classify();
+    assert(LowerMask <= UpperMask && "Range is nan-only.");
+    for (uint32_t I = LowerMask; I <= UpperMask; I <<= 1)
+      Mask |= I;
+  }
```
Comment (jayfoad) [inline llvm/lib/IR/ConstantFPRange.cpp:195]: You can do this without a loop: `Mask |= (UpperMask << 1) - LowerMask` (with a comment)

**Pull request merged**